### PR TITLE
[script.skinshortcuts] v1.1.0

### PR DESCRIPTION
**Pull request by mikesilvo164**
### Description

Python 3 version of Skin Shortcuts
### Checklist:

- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by basrieter** [comment]
I am confused by 2 devs requesting PR's for the same add-on? https://github.com/xbmc/repo-scripts/pull/1317

**Comment by mikesilvo164** [comment]
I have no idea what that other person is doing. I made the script python 3 compatible with help from the original script creator.

**Comment by sualfred** [comment]
@basrieter 

The other commit is just a fork that is flying around since years. This one is the only one which is ported to Python3 based on the original upstream of BigNoid.

File: script.skinshortcuts/addon.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		<description lang="no">Tillegg for enkel administrering og opplisting av bruker-snarveier i tema (krever støtte i tema)</description>
+		<description lang="pl">Wtyczka dla skór, która zapewnia proste zarządzanie i tworzenie list skrótów użytkownika (wymaga wsparcia skóry)</description>
+		<description lang="pt">Um add-on que permite que os temas tenham recursos simples para gerir e listar atalhos de utilizadores (é necessário suporte do tema)</description>
+		<description lang="pt_BR">Add-on para que as skins possam prover gerenciamento de lista de atalhos para o menu principal (requer suporte da skin)</description>
+		<description lang="ru">Дополнение для обложек, предоставляющее простой способ управления и список пользовательских ярлыков (требуется поддержка обложкой)</description>
+		<description lang="sk">Doplnok poskytujúci jednoduchý manažment zoznamu užívateľských odkazov (vyžaduje podporu v doplnku pre vzhľad)</description>
+		<description lang="sv">Skinntillägg som tillhandahåller enkel hantering och översikt av användargenvägar (om skinnet stöder det)</description>
+		<description lang="zh">插件为皮肤提供了简单的管理和用户连接方式列表 (需要皮肤支持)</description>
```
Comment (sualfred) [inline script.skinshortcuts/addon.xml]: please use 

> en_GB

for example

File: script.skinshortcuts/addon.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		<language></language>
+		<platform>all</platform>
+		<license>GNU General Public License, v2</license>
```
Comment (sualfred) [inline script.skinshortcuts/addon.xml]: a spdx identifier (https://spdx.org/ids-how) is better

File: script.skinshortcuts/addon.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<addon id="script.skinshortcuts" name="Skin Shortcuts" version="1.1.0" provider-name="BigNoid, marcelveldt, SiLVO">
+	<requires>
+		<import addon="xbmc.python" version="3.0.0"/>
+		<import addon="script.module.simplejson" version="3.3.0"/>
```
Comment (sualfred) [inline script.skinshortcuts/addon.xml]: No need for script.module.simplejson. It's not used anywhere in the code and you already switched to the included "json" module.

File: script.skinshortcuts/resources/lib/datafunctions.py [hunk 4]
```diff
@@ -0,0 +1,1372 @@
...
+            # Not an 'ActivateWindow' - return the onclick
+            return onclick
+
+    def upgradeAction( self, action ):
+        # This function looks for actions used in a previous version of Kodi, and upgrades them to the current action
+
+        if not action.lower().startswith( "activatewindow(" ): return action
+
+        # Isengard + earlier music addons
+        if int( KODIVERSION ) <= 15:
```
Comment (sualfred) [inline script.skinshortcuts/resources/lib/datafunctions.py]: No need for this kind of checks anymore

File: script.skinshortcuts/resources/lib/datafunctions.py [hunk 5]
```diff
@@ -0,0 +1,1372 @@
...
+                elif action.lower().startswith( "activatewindow(10502" ) and xbmc.getCondVisibility( "!Library.HasContent(Music)" ):
+                    return self.buildReplacementMusicAddonAction( action, "10501" )
+                elif action.lower().startswith( "activatewindow(musicfiles" ) and xbmc.getCondVisibility( "Library.HasContent(Music)" ):
+                    return self.buildReplacementMusicAddonAction( action, "MusicLibrary" )
+                elif action.lower().startswith( "activatewindow(10501" ) and xbmc.getCondVisibility( "Library.HasContent(Music)" ):
+                    return self.buildReplacementMusicAddonAction( action, "10502" )
+
+
+        # Jarvis + later music windows
+        if action.lower() == "activatewindow(musicfiles)" and int( KODIVERSION ) >= 16:
```
Comment (sualfred) [inline script.skinshortcuts/resources/lib/datafunctions.py]: No need for this kind of checks anymore

File: script.skinshortcuts/resources/lib/datafunctions.py [hunk 6]
```diff
@@ -0,0 +1,1372 @@
...
+            return "ActivateWindow(Music,Files,Return)"
+
+        if action.lower().startswith("activatewindow(musiclibrary") and int( KODIVERSION ) >= 16:
```
Comment (sualfred) [inline script.skinshortcuts/resources/lib/datafunctions.py]: No need for this kind of checks anymore

File: script.skinshortcuts/resources/lib/nodefunctions.py [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+HEX_REXP = re.compile('&#x([\da-fA-F]+);')
+
+REPLACE1_REXP = re.compile(r'[\']+')
+REPLACE2_REXP = re.compile(r'[^-a-z0-9]+')
+REMOVE_REXP = re.compile('-{2,}')
+
+def log(txt):
+    if ADDON.getSetting( "enable_logging" ) == "true":
+        if not isinstance (txt,str):
+            txt = txt.decode('utf-8')
```
Comment (sualfred) [inline script.skinshortcuts/resources/lib/nodefunctions.py]: no need for decoding in Py3

File: script.skinshortcuts/resources/lib/unicodeutils.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# coding=utf-8
+def try_decode(text, encoding="utf-8"):
```
Comment (sualfred) [inline script.skinshortcuts/resources/lib/unicodeutils.py]: I assume this is a Py2 leftover?

**Comment by mikesilvo164** [comment]
@sualfred IMHO everything in upgradeAction is needed in case someone decides to upgrade from a very old version of Kodi. The shortcuts would still work then.

File: script.skinshortcuts/addon.xml (hunk 1 above)
Comment (mikesilvo164) [reply script.skinshortcuts/addon.xml]: I assume you mean switching all the description lang= to the matching ending language folder names or just the en one?

**Comment by sualfred** [comment]
@mikesilvo164 
KODIVERSION is "19" and this conditions will never be True. Since this version will never run on < Matrix you just have the keep the upgrade part which is required for Matrix. The lines for Isengard, Jarvis, etc, can be dropped.

Please check your addon.xml for the error (website) and take a look at the result of travis. There are a lot of refactored things you should change.

**Comment by mikesilvo164** [comment]
@sualfred Done 👍

File: script.skinshortcuts/changelog.txt [hunk 9]
```diff
@@ -0,0 +1,225 @@
+v1.0.13
```
Comment (enen92) [inline script.skinshortcuts/changelog.txt]: can you drop this file from submission please?

File: script.skinshortcuts/default.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+ADDONID      = ADDON.getAddonInfo('id')
+ADDONVERSION = ADDON.getAddonInfo('version')
+LANGUAGE     = ADDON.getLocalizedString
+CWD          = ADDON.getAddonInfo('path')
+ADDONNAME    = ADDON.getAddonInfo('name')
+RESOURCE     = xbmc.translatePath(os.path.join(CWD, 'resources', 'lib'))
+DATAPATH     = os.path.join(xbmc.translatePath("special://profile/"), "addon_data", ADDONID)
+MASTERPATH   = os.path.join(xbmc.translatePath("special://masterprofile/"), "addon_data", ADDONID)
+
+sys.path.append(RESOURCE)
```
Comment (enen92) [inline script.skinshortcuts/default.py]: please avoid appending to sys.path, kodi interpreter already add your depends to the path of this addon.
Use absolute imports:

`from resources.lib import x,y,z`

File: script.skinshortcuts/default.py [hunk 11]
```diff
@@ -0,0 +1,397 @@
+# coding=utf-8
```
Comment (enen92) [inline script.skinshortcuts/default.py]: It's a good practice to keep this file as small as possible as it's the only file not compiled to bytecode (as it's the entrypoint). Please move the codebase to `/resources/lib/` and in this file just do something like:

```python
import sys
from resources.lib import myaddon

myaddon.run(sys.argv)
```

File: script.skinshortcuts/resources/docs/FAQ.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
+# End User FAQ
```
Comment (enen92) [inline script.skinshortcuts/resources/docs/FAQ.md]: please avoid submitting the docs folder. Add it to .gitattributes

File: script.skinshortcuts/resources/lib/datafunctions.py [hunk 13]
```diff
@@ -0,0 +1,1339 @@
...
+        else:
+            # Not an 'ActivateWindow' - return the onclick
+            return onclick
+
+    def upgradeAction( self, action ):
+        # This function looks for actions used in a previous version of Kodi, and upgrades them to the current action
+
+        if not action.lower().startswith("activatewindow("): return action
+
+        # Jarvis + later music windows
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/datafunctions.py]: maybe time for a bit of cleanup isn't it? :)

File: script.skinshortcuts/resources/lib/gui.py [hunk 14]
```diff
@@ -0,0 +1,2417 @@
...
+import urllib.request, urllib.parse, urllib.error
+import _thread as thread
+
+import datafunctions
+DATA = datafunctions.DataFunctions()
+
+import library
+LIBRARY = library.LibraryFunctions()
+
+ADDON        = sys.modules[ "__main__" ].ADDON
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/gui.py]: move all this to a common file and get values from there

File: script.skinshortcuts/resources/lib/simpleeval.py [hunk 15]
```diff
@@ -0,0 +1,313 @@
+'''
+SimpleEval - (C) 2013/2015 Daniel Fairhead
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/simpleeval.py]: please move this to a script.module

Comment (enen92) [reply script.skinshortcuts/resources/lib/simpleeval.py]: license is also different from the one of your main addon.

File: script.skinshortcuts/resources/lib/datafunctions.py (hunk 13 above)
Comment (sualfred) [reply script.skinshortcuts/resources/lib/datafunctions.py]: @enen92 
He already did. This part is required if the user upgraded from a very old version with old skinshortcuts data. This automatically updates the old actions to valid ones.

Comment (enen92) [reply script.skinshortcuts/resources/lib/datafunctions.py]: ok

File: script.skinshortcuts/resources/lib/simpleeval.py (hunk 15 above)
Comment (mikesilvo164) [reply script.skinshortcuts/resources/lib/simpleeval.py]: Is this really necessary? Seems silly to add a seperate module to maintain.

File: script.skinshortcuts/resources/lib/gui.py (hunk 14 above)
Comment (mikesilvo164) [reply script.skinshortcuts/resources/lib/gui.py]: Can you please clarify this?

File: script.skinshortcuts/resources/docs/FAQ.md (hunk 12 above)
Comment (mikesilvo164) [reply script.skinshortcuts/resources/docs/FAQ.md]: Not exactly sure how to do this...
```
/docs               export-ignore
```
Is that correct?

Comment (mikesilvo164) [reply script.skinshortcuts/resources/docs/FAQ.md]: Since no response I just deleted the docs folder

**Comment by mikesilvo164** [comment]
The build is fine don't know why travis fails like this sometimes 🤷‍♂

File: script.skinshortcuts/resources/lib/gui.py [hunk 16]
```diff
@@ -0,0 +1,2417 @@
...
+
+        # If either localLabel[ 2 ] starts with a $, ask Kodi to parse it for us
+        if localLabel[ 2 ].startswith( "$" ):
+            localLabel[ 2 ] = xbmc.getInfoLabel( localLabel[ 2 ] )
+        if localLabel2[ 2 ].startswith( "$" ):
+            localLabel2[ 2 ] = xbmc.getInfoLabel( localLabel2[ 2 ] )
+
+        # Create the list item
+        listitem = xbmcgui.ListItem(label=localLabel[2], label2 = localLabel2[2])
+        listitem.setArt({'icon': xbmc.getInfoLabel(icon)})
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/gui.py:335]: make this a single call in the future

File: script.skinshortcuts/resources/lib/gui.py [hunk 17]
```diff
@@ -0,0 +1,2417 @@
...
+        listitem.setArt({'thumb': xbmc.getInfoLabel(thumb)})
+        listitem.setProperty( "localizedString", localLabel[0] )
+        listitem.setProperty( "icon", icon )
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/gui.py:338]: is this needed?

File: script.skinshortcuts/resources/lib/gui.py [hunk 18]
```diff
@@ -0,0 +1,2417 @@
...
+                newItem.setProperty( "action-show", item.getProperty( "action-show" ) )
+                newItem.setProperty( "action-play", item.getProperty( "action-play" ) )
+                newItem.setProperty( "action-party", item.getProperty( "action-party" ) )
+            self.getControl( 111 ).addItem( newItem )
+        self.getControl( 101 ).setLabel( label + " (%s)" %self.getControl( 111 ).size() )
+
+    def _duplicate_listitem( self, listitem, originallistitem = None ):
+        # Create a copy of an existing listitem
+        listitemCopy = xbmcgui.ListItem(label=listitem.getLabel(), label2=listitem.getLabel2())
+        listitem.setArt({'icon': listitem.getProperty("icon")})
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/gui.py:2156]: same

File: script.skinshortcuts/resources/lib/library.py [hunk 19]
```diff
@@ -0,0 +1,2163 @@
...
+
+        # If the skin doesn't have the icon, replace it with DefaultShortcut.png
+        if ( not displayIcon or not xbmc.skinHasImage( displayIcon ) ) and not iconIsVar:
+            if not usedDefaultThumbAsIcon:
+                displayIcon = "DefaultShortcut.png"
+
+        # Build listitem
+        if thumbnail is not None:
+            listitem = xbmcgui.ListItem(label=displayLabel, label2=displayLabel2)
+            listitem.setArt({'icon': displayIcon})
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/library.py:619]: same

File: script.skinshortcuts/resources/lib/library.py [hunk 20]
```diff
@@ -0,0 +1,2163 @@
...
+        else:
+            self.getControl(5).setVisible(False)
+        self.getControl(1).setLabel(self.windowtitle)
+
+        # Set Cancel label (Kodi 17+)
+        self.getControl(7).setLabel(xbmc.getLocalizedString(222))
+
+        for item in self.listing :
+            listitem = xbmcgui.ListItem(label=item.getLabel(), label2=item.getLabel2())
+            listitem.setArt({'icon': item.getProperty("icon")})
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/library.py:2139]: same, the less calls to methods the better

File: script.skinshortcuts/resources/lib/nodefunctions.py [hunk 21]
```diff
@@ -0,0 +1,596 @@
...
+        # Build the list of all properties to save
+        saveData = []
+        for saveGroup in allProps:
+            for saveLabelID in allProps[ saveGroup ]:
+                for saveProperty in allProps[ saveGroup ][ saveLabelID ]:
+                    saveData.append( [ saveGroup, saveLabelID, saveProperty, allProps[ saveGroup ][ saveLabelID ][ saveProperty ] ] )
+
+        # Save the new properties
+        try:
+            f = xbmcvfs.File(os.path.join(DATAPATH, xbmc.getSkinDir() + ".properties"), 'w')
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/nodefunctions.py:518]: this supports the `with` statement in Kodi 19

File: script.skinshortcuts/resources/lib/nodefunctions.py [hunk 22]
```diff
@@ -0,0 +1,596 @@
...
+        else:
+            self.getControl(5).setVisible(False)
+        self.getControl(1).setLabel(self.windowtitle)
+
+        # Set Cancel label (Kodi 17+)
+        self.getControl(7).setLabel(xbmc.getLocalizedString(222))
+
+        for item in self.listing :
+            listitem = xbmcgui.ListItem(label=item.getLabel(), label2=item.getLabel2())
+            listitem.setArt({'icon': item.getProperty("icon")})
```
Comment (enen92) [inline script.skinshortcuts/resources/lib/nodefunctions.py:571]: same

File: script.skinshortcuts/resources/lib/simpleeval.py (hunk 15 above)
Comment (enen92) [reply script.skinshortcuts/resources/lib/simpleeval.py:2]: Yeah that's the rule but you might do it later, as long as you actually do it :)

**Comment by mikesilvo164** [comment]
@enen92 The icon stuff is needed in case the skinner/user overrides the icon or am I missing something?

File: script.skinshortcuts/resources/lib/nodefunctions.py (hunk 21 above)
Comment (mikesilvo164) [reply script.skinshortcuts/resources/lib/nodefunctions.py:518]: Could you explain this in more detail please. I am kind of blindly stumbling through Python development...

**Comment by enen92** [comment]
My comment was about the icon property. As the icon is already defined in the artmap I am wondering if the property is needed. But no mandatory change about that

**Comment by basrieter** [comment]
Merging this. Please try to look into the suggestions of @enen92 for a next version.

**Pull request merged**